src/native/libs/System.Security.Cryptography.Native/pal_evp_pkey.c

```diff
@@ -3,92 +3,114 @@
+struct EvpPKeyExtraHandle_st
 {
-    assert(currentKey != NULL);
-
-    ERR_clear_error();
-
-    int currentAlgId = EVP_PKEY_get_base_id(currentKey);
-
-    if (algId != NID_undef && algId != currentAlgId)
-    {
-        ERR_put_error(ERR_LIB_EVP, 0, EVP_R_DIFFERENT_KEY_TYPES, __FILE__, __LINE__);
-        return NULL;
-    }
-
-    EVP_PKEY* newKey = EVP_PKEY_new();
-
-    if (newKey == NULL)
-    {
-        return NULL;
-    }
+    int references;
+    OSSL_LIB_CTX* libCtx;
+    OSSL_PROVIDER* prov;
+};
 
-    bool success = true;
+typedef struct EvpPKeyExtraHandle_st EvpPKeyExtraHandle;
 
-    if (currentAlgId == EVP_PKEY_RSA)
-    {
-        const RSA* rsa = EVP_PKEY_get0_RSA(currentKey);
+static void CryptoNative_EvpPkeyExtraHandleDestroy(EvpPKeyExtraHandle* handle)
+{
+    assert(handle->references >= 1);
+    assert(handle->prov != NULL);
+    assert(handle->libCtx != NULL);
 
-#pragma clang diagnostic push
-#pragma clang diagnostic ignored "-Wcast-qual"
-        if (rsa == NULL || !EVP_PKEY_set1_RSA(newKey, (RSA*)rsa))
-#pragma clang diagnostic pop
-        {
-            success = false;
-        }
-    }
-    else
-    {
-        ERR_put_error(ERR_LIB_EVP, 0, EVP_R_UNSUPPORTED_ALGORITHM, __FILE__, __LINE__);
-        success = false;
-    }
+    handle->references--;
 
-    if (!success)
+    if (handle->references == 0)
     {
-        EVP_PKEY_free(newKey);
-        newKey = NULL;
+        OSSL_PROVIDER_unload(handle->prov);
+        OSSL_LIB_CTX_free(handle->libCtx);
+        free(handle);
     }
+}
+#endif
 
-    return newKey;
+EVP_PKEY* CryptoNative_EvpPkeyCreate(void)
+{
+    ERR_clear_error();
+    return EVP_PKEY_new();
 }
 
-void CryptoNative_EvpPkeyDestroy(EVP_PKEY* pkey)
+void CryptoNative_EvpPkeyDestroy(EVP_PKEY* pkey, void* extraHandle)
 {
     if (pkey != NULL)
     {
         EVP_PKEY_free(pkey);
     }
+
+#ifdef NEED_OPENSSL_3_0
+    if (extraHandle != NULL)
+    {
+        EvpPKeyExtraHandle* extra = (EvpPKeyExtraHandle*)extraHandle;
+        CryptoNative_EvpPkeyExtraHandleDestroy(extra);
+    }
+#else
+    (void)extraHandle;
+    assert(extraHandle == NULL);
+#endif
 }
 
-int32_t CryptoNative_EvpPKeySize(EVP_PKEY* pkey)
+int32_t CryptoNative_EvpPKeyBits(EVP_PKEY* pkey)
 {
     // This function is not expected to populate the error queue with
     // any errors, but it's technically possible that an external
     // ENGINE or OSSL_PROVIDER populate the queue in their implementation,
     // but the calling code does not check for one.
     assert(pkey != NULL);
-    return EVP_PKEY_get_size(pkey);
+    return EVP_PKEY_get_bits(pkey);
 }
 
-int32_t CryptoNative_UpRefEvpPkey(EVP_PKEY* pkey)
+int32_t CryptoNative_UpRefEvpPkey(EVP_PKEY* pkey, void* extraHandle)
 {
     if (!pkey)
     {
         return 0;
     }
 
+#ifdef NEED_OPENSSL_3_0
+    EvpPKeyExtraHandle* extra = (EvpPKeyExtraHandle*)extraHandle;
+
+    if (extra != NULL)
+    {
+        extra->references++;
```

Comment:
Do we need atomics here? In Android where we do manual reference counting, we use an atomic int. 

https://github.com/dotnet/runtime/blob/9565550164fd63f1612141d221e04747d9742d4b/src/native/libs/System.Security.Cryptography.Native.Android/pal_eckey.h#L13

I don't think you can use that Android code as-is, but I am curious if we have concerns about two different threads increasing the reference count for the extra handle.